cpp/dolfinx/function/Function.h

```diff
@@ -89,7 +95,21 @@ class Function
   }
 
   /// Move assignment
-  Function& operator=(Function&& v) = default;
+  Function& operator=(Function&& v)
+  {
+    if (this != &v)
```

Comment:
Why are you checking this? Self-move-assignment will be most likely undefined behaviour in all std lib moves below anyway.

Reply:
It's a standard idiom of C++ that assignment operators should be safe with respect to self-assignment. In the case of move assignment there's no consensus: cppreference has snippets where it is used [e.g. [in a note here](https://en.cppreference.com/w/cpp/utility/exchange)] whereas isocpp mentions that it shouldn't be an issue. I thought I added it, "better safe than sorry".

Reply:
I'm not super keen on the check. One needs to do really weird (and bad) things to trigger it.

Reply:
It feels very weird to me to handle this, and makes the code ugly as well. Maybe just add and assert() if you want to be safe.

Reply:
I removed the check.

Reply:
Consider some function that consumes a Function and either returns a new one or just forwards the given Function if no new Function is needed for whatever reason. Assign the result to the input, .e.g.
```
Function myfunc;
...
myfunc = modify_or_new(myfunc, other_args);
```
I believe, that there might be valid reasons for such or at least similar a constructions. A "practical" example might be
to append elements to a vector with std::accumulate instead of using std::transform. I have working code that does that so I would say that std containers support self-move assignment.

Reply:
With the latest version of the code self-move-assignment of Function is possible as far as all std lib implementations will allow it.

Reply:
Right, I forgot about the code that followed the check. Should behave correctly as it is. Beyond that it's probably just a matter of taste.